### PR TITLE
fix: error responses

### LINKED ISSUE #15: fix: error responses
https://walletconnect.slack.com/archives/C044SKFKELR/p1691584279490409?thread_ts=1691503019.962459&cid=C044SKFKELR

### PR DESCRIPTION
**Pull request by chris13524**
# Description

Sends JSON-RPC error responses when an error is encountered wherever possible rather than not responding at all.

Remaining work:
- [x] Use standard error payload instead of String
- [x] Obscure server error reason
- [ ] ~~Update tests to test client error reason~~ Won't do for now
- [ ] ~~Can more error cases be returned to the client instead of dropped~~ Yes, but not worth it for now
- [ ] ~~Update specs to include error codes~~ Won't do for now; no specific error codes are defined

Resolves #15 

## How Has This Been Tested?

Updated automated tests

## Due Diligence

* [ ] Breaking change
* [ ] Requires a documentation update
* [ ] Requires a e2e/integration test update

### CODE REVIEW

**Review by geekbrother — APPROVED** [review]

File: src/rpc.rs [hunk 1]
```diff
@@ -42,7 +42,7 @@ pub struct JsonRpcRequest<T> {
 
 impl<T> JsonRpcRequest<T> {
     pub fn new(method: &str, params: T) -> Self {
-        JsonRpcRequest {
+        Self {
```
Comment (nopestack) [inline src/rpc.rs:45]: minor style q: do we have a preference as a team to use Self vs the type name itself?

File: src/rpc.rs [hunk 2]
```diff
@@ -60,14 +60,39 @@ pub struct JsonRpcResponse<T> {
...
+        }
+    }
+}
+
+// https://www.jsonrpc.org/specification#error_object
+#[derive(Debug, Serialize, Deserialize)]
+pub struct JsonRpcError {
+    pub code: i16,
+    pub message: String,
+    // pub data: Value,
```
Comment (nopestack) [inline src/rpc.rs:93]: why was this commented?

File: src/rpc.rs (hunk 1 above)
Comment (chris13524) [reply src/rpc.rs:45]: No need to enforce. I liked Self in this particular circumstance

File: src/services/public_http_server/handlers/relay_webhook/handlers/notify_delete.rs [hunk 3]
```diff
@@ -77,115 +85,135 @@ pub async fn handle(msg: RelayIncomingMessage, state: &AppState) -> Result<(), R
...
+        project_client_id: DecodedClientId,
+    ) -> Result<
+        (
+            ResponseAuth,
+            Vec<(SubscriptionWatcherQuery, Vec<NotifyServerSubscription>)>,
+        ),
+        RelayMessageError,
+    > {
+        info!("req.id: {}", req.id);
+        info!("req.jsonrpc: {}", req.jsonrpc); // TODO verify this
```
Comment (nopestack) [inline src/services/public_http_server/handlers/relay_webhook/handlers/notify_delete.rs:112]: Can you elaborate on what exactly needs to verify in these two lines?

Comment (chris13524) [reply src/services/public_http_server/handlers/relay_webhook/handlers/notify_delete.rs:112]: Just that `jsonrpc` matches `2.0` and the `method` matches the endpoint

File: src/rpc.rs (hunk 2 above)
Comment (chris13524) [reply src/rpc.rs:93]: JSON-RPC errors have a data field which can have more custom data. But we don't need it right now

File: src/services/public_http_server/handlers/relay_webhook/handlers/notify_subscribe.rs [hunk 4]
```diff
@@ -82,195 +91,215 @@ pub async fn handle(msg: RelayIncomingMessage, state: &AppState) -> Result<(), R
...
-            WebhookNotificationEvent::Subscribed,
-            // } else {
-            // WebhookNotificationEvent::Updated
-            // },
-            account.as_ref(),
+        // TODO do in same transaction as upsert_subscriber()
+        state
+            .notify_webhook(
+                project.project_id.as_ref(),
+                // TODO uncomment when `WebhookNotificationEvent::Updated` exists
```
Comment (nopestack) [inline src/services/public_http_server/handlers/relay_webhook/handlers/notify_subscribe.rs:257]: I'm curious about this event, does it exists already?

Comment (chris13524) [reply src/services/public_http_server/handlers/relay_webhook/handlers/notify_subscribe.rs:257]: No it does not

**Review by nopestack — APPROVED** [review]

**Pull request merged**